src/dotnet/UI.Blazor/Components/ContextMenu/BlazorContextMenu/Components/ContextMenuBase.cs

```diff
@@ -0,0 +1,263 @@
+        }
+    }
+    protected string ListClassCalc {
+        get {
+            var template = Settings.GetTemplate(GetActiveTemplate());
+            return Helpers.AppendCssClasses((OverrideDefaultListCssClass ?? template.DefaultCssOverrides.MenuListCssClass),
+                (ListCssClass ?? Settings.GetTemplate(GetActiveTemplate()).MenuListCssClass));
+        }
+    }
+
+    protected (string showingClass, string hiddenClass) GetAnimationClasses(Animation animation)
+    {
+        switch (animation)
+        {
+        case BlazorContextMenu.Animation.None:
+            return ("", "");
+        case BlazorContextMenu.Animation.FadeIn:
+            return ("blazor-context-menu__animations--fadeIn-shown", "blazor-context-menu__animations--fadeIn");
+        case BlazorContextMenu.Animation.Grow:
+            return ("blazor-context-menu__animations--grow-shown", "blazor-context-menu__animations--grow");
+        case BlazorContextMenu.Animation.Slide:
+            return ("blazor-context-menu__animations--slide-shown", "blazor-context-menu__animations--slide");
+        case BlazorContextMenu.Animation.Zoom:
+            return ("blazor-context-menu__animations--zoom-shown", "blazor-context-menu__animations--zoom");
+        default:
+            throw new Exception("Animation not supported");
+        }
+    }
+
+    protected override void OnInitialized()
+    {
+        base.OnInitialized();
+        if (string.IsNullOrEmpty(Id))
+            throw new ArgumentNullException(nameof(Id));
+        ContextMenuStorage.Register(this);
+    }
+
+    protected override async Task OnAfterRenderAsync(bool firstRender)
+    {
+        if (!ContextMenuHandler.ReferencePassedToJs) {
+            await JSRuntime.InvokeAsync<object>("blazorContextMenu.SetMenuHandlerReference", DotNetObjectReference.Create(ContextMenuHandler));
+            ContextMenuHandler.ReferencePassedToJs = true;
+        }
+    }
+
+    public override void Dispose()
+    {
+        base.Dispose();
+        ContextMenuStorage.Unregister(this);
+    }
+
+    internal async Task Show(string x, string y, string? targetId = null, ContextMenuTrigger? trigger = null)
+    {
+        if (trigger is null) {
+            var rootMenu = MenuTreeTraverser.GetRootContextMenu(this);
+            trigger = rootMenu?.GetTrigger();
+        }
+
+        if (trigger != null)
+            Data = trigger.Data;
+
+        if (OnAppearing.HasDelegate) {
+            var eventArgs = new MenuAppearingEventArgs(Id, targetId, x, y, trigger, Data);
+            await OnAppearing.InvokeAsync(eventArgs);
+            x = eventArgs.X;
+            y = eventArgs.Y;
+            if (eventArgs.PreventShow)
+                return;
+        }
+
+        IsShowing = true;
+        X = x;
+        Y = y;
+        TargetId = targetId;
+        Trigger = trigger;
+        await InvokeAsync(() => StateHasChanged());
```

Comment:
It is kinda suspicious this method can be called from non-UI thread, even though it runs just UI-related logic, + worse, it calls event handlers.

If you need to route the work to the UI thread, it seems better to do this when you invoke this method, but not inside it.

Reply:
removed